addon/-private/system/model/internal-model-class.js

```diff
@@ -0,0 +1,615 @@
+import Ember from 'ember';
+import {
+  relationshipFromMeta
+} from 'ember-data/-private/system/relationship-meta';
+import { assert, warn } from "ember-data/-private/debug";
+
+const {
+  get,
+  Map,
+  A
+} = Ember;
+
+function findPossibleInverses(type, inverseType, name, relationshipsSoFar) {
+  let possibleRelationships = relationshipsSoFar || [];
+  let relationshipMap = get(inverseType, 'relationships');
+
+  if (!relationshipMap) { return possibleRelationships; }
+
+  let relationships = relationshipMap.get(type.modelName).filter(relationship => {
+    let optionsForRelationship = inverseType.metaForProperty(relationship.name).options;
+
+    if (!optionsForRelationship.inverse) {
+      return true;
+    }
+
+    return name === optionsForRelationship.inverse;
+  });
+
+  if (relationships) {
+    possibleRelationships.push.apply(possibleRelationships, relationships);
+  }
+
+  //Recurse to support polymorphism
+  if (type.superclass) {
+    findPossibleInverses(type.superclass, inverseType, name, possibleRelationships);
+  }
+
+  return possibleRelationships;
+}
+
+
+export default class InternalModelClass {
+  constructor(modelClass) {
+    let attributes;
+    let relationships;
+    let descriptors = get(modelClass, '_computedProperties');
```

Comment:
I experimented with doing it lazily, but in nearly all cases currently we access additional information almost immediately following lookup.